### PR TITLE
Create lifecycle test against AWS mocks

### PR DESCRIPTION
**Pull request by justinsb**
We create a cluster, run kops update, then run kops update again.  We
expect no changes on the second update.

### CODE REVIEW

**Comment by justinsb** [comment]
Note this is currently expected to fail with:

```
        lifecycle_integration_test.go:127: Target had changes after executing: Will modify resources:
                  DHCPOptions/minimal.example.com
                        Shared                   <nil> -> false
                        Tags                     <nil> -> {Name: minimal.example.com, KubernetesCluster: minimal.example.com, kubernetes.io/cluster/minimal.example.com: owned}

                  InternetGateway/minimal.example.com
                        Tags                     <nil> -> {Name: minimal.example.com, KubernetesCluster: minimal.example.com, kubernetes.io/cluster/minimal.example.com: owned}

                  RouteTable/minimal.example.com
                        Shared                   <nil> -> false
                        Tags                     <nil> -> {Name: minimal.example.com, KubernetesCluster: minimal.example.com, kubernetes.io/cluster/minimal.example.com: owned}
```

File: cloudmock/aws/mockautoscaling/group.go [hunk 1]
```diff
@@ -18,52 +18,129 @@ package mockautoscaling
...
+	g := &autoscaling.Group{
 		AutoScalingGroupName: input.AutoScalingGroupName,
-		MinSize:              input.MinSize,
-		MaxSize:              input.MaxSize,
-		Instances:            []*autoscaling.Instance{},
+		AvailabilityZones:    input.AvailabilityZones,
+		CreatedTime:          &createdTime,
+		DefaultCooldown:      input.DefaultCooldown,
+		DesiredCapacity:      input.DesiredCapacity,
+		// EnabledMetrics:          input.EnabledMetrics,
```
Comment (chrislovecnm) [inline cloudmock/aws/mockautoscaling/group.go]: Don’t we have this now in our code?

Comment (chrislovecnm) [reply cloudmock/aws/mockautoscaling/group.go]: Nit pick to be honest

File: cloudmock/aws/mockautoscaling/group.go [hunk 2]
```diff
@@ -76,10 +153,44 @@ func (m *MockAutoscaling) TerminateInstanceInAutoScalingGroup(input *autoscaling
...
+				}, nil
 			}
 		}
 	}
 
 	return nil, fmt.Errorf("Instance not found")
 }
+
+func (m *MockAutoscaling) DescribeAutoScalingGroupsWithContext(aws.Context, *autoscaling.DescribeAutoScalingGroupsInput, ...request.Option) (*autoscaling.DescribeAutoScalingGroupsOutput, error) {
+	glog.Fatalf("Not implemented")
```
Comment (chrislovecnm) [inline cloudmock/aws/mockautoscaling/group.go]: You always say to not use duplicate log messages. When not implemented. I think flatal returns a err. Might as well just return an err here.

File: cloudmock/aws/mockautoscaling/group.go [hunk 3]
```diff
@@ -18,52 +18,129 @@ package mockautoscaling
...
 )
 
 func (m *MockAutoscaling) AttachInstances(input *autoscaling.AttachInstancesInput) (*autoscaling.AttachInstancesOutput, error) {
-	for _, group := range m.Groups {
-		if aws.StringValue(group.AutoScalingGroupName) == aws.StringValue(input.AutoScalingGroupName) {
-			for _, instanceID := range input.InstanceIds {
-				group.Instances = append(group.Instances, &autoscaling.Instance{InstanceId: instanceID})
-			}
-		}
+	m.mutex.Lock()
```
Comment (chrislovecnm) [inline cloudmock/aws/mockautoscaling/group.go]: Being picky but do we need func comments how this works for mocking. This is a lot of useful code, only if developers understand how to reuse it for testing.

File: cloudmock/aws/mockautoscaling/group.go [hunk 4]
```diff
@@ -18,52 +18,129 @@ package mockautoscaling
...
+	defer m.mutex.Unlock()
+
+	glog.Infof("EnableMetricsCollection: %v", request)
+
+	g := m.Groups[*request.AutoScalingGroupName]
+	if g == nil {
+		return nil, fmt.Errorf("AutoScalingGroup not found")
+	}
+
+	metrics := make(map[string]*autoscaling.EnabledMetric)
```
Comment (chrislovecnm) [inline cloudmock/aws/mockautoscaling/group.go]: Looks like the same code as in groups? Maybe it is different interfaces but the same business logic.

Comment (chrislovecnm) [reply cloudmock/aws/mockautoscaling/group.go]: Never mind the rebase messed me ups ignore the comment,

File: cloudmock/aws/mockec2/routetable.go [hunk 5]
```diff
@@ -34,19 +51,138 @@ func (m *MockEC2) DescribeRouteTablesWithContext(aws.Context, *ec2.DescribeRoute
...
 	}
 
 	response := &ec2.DescribeRouteTablesOutput{}
+
 	for _, rt := range m.RouteTables {
-		response.RouteTables = append(response.RouteTables, rt)
+		allFiltersMatch := true
+		for _, filter := range request.Filters {
+			match := false
+			switch *filter.Name {
```
Comment (chrislovecnm) [inline cloudmock/aws/mockec2/routetable.go]: Should we use the aws utility aws.String? Get the string safely.

File: cloudmock/aws/mockec2/routetable.go [hunk 6]
```diff
@@ -34,19 +51,138 @@ func (m *MockEC2) DescribeRouteTablesWithContext(aws.Context, *ec2.DescribeRoute
...
+
+	copy := *rt
+	response := &ec2.CreateRouteTableOutput{
+		RouteTable: &copy,
+	}
+	return response, nil
+}
+
+func (m *MockEC2) CreateRouteTableWithContext(aws.Context, *ec2.CreateRouteTableInput, ...request.Option) (*ec2.CreateRouteTableOutput, error) {
+	panic("Not implemented")
```
Comment (chrislovecnm) [inline cloudmock/aws/mockec2/routetable.go]: You switched from glog to panics

File: cloudmock/aws/mockec2/routetable.go [hunk 7]
```diff
@@ -34,19 +51,138 @@ func (m *MockEC2) DescribeRouteTablesWithContext(aws.Context, *ec2.DescribeRoute
...
+
+func (m *MockEC2) CreateRoute(request *ec2.CreateRouteInput) (*ec2.CreateRouteOutput, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	glog.Infof("CreateRoute: %v", request)
+
+	if request.DryRun != nil {
+		glog.Fatalf("DryRun")
+	}
```
Comment (chrislovecnm) [inline cloudmock/aws/mockec2/routetable.go]: Does you mean to panic here?

File: cloudmock/aws/mockec2/volumes.go [hunk 8]
```diff
@@ -69,6 +127,10 @@ func (m *MockEC2) DescribeVolumesWithContext(aws.Context, *ec2.DescribeVolumesIn
 func (m *MockEC2) DescribeVolumes(request *ec2.DescribeVolumesInput) (*ec2.DescribeVolumesOutput, error) {
 	glog.Infof("DescribeVolumes: %v", request)
 
+	if request.VolumeIds != nil {
+		glog.Fatalf("VolumeIds")
```
Comment (chrislovecnm) [inline cloudmock/aws/mockec2/volumes.go]: Message is not actionable

File: cloudmock/aws/mockautoscaling/group.go (hunk 2 above)
Comment (justinsb) [reply cloudmock/aws/mockautoscaling/group.go]: glog.Fatalf does a stack-trace, so we're OK :-)

File: cloudmock/aws/mockec2/routetable.go (hunk 5 above)
Comment (justinsb) [reply cloudmock/aws/mockec2/routetable.go]: Probably yes, but we aren't really aiming for 100% fidelity and reliability.

File: cloudmock/aws/mockec2/volumes.go (hunk 8 above)
Comment (justinsb) [reply cloudmock/aws/mockec2/volumes.go]: It is with a stack trace :-)

**Pull request merged**